js/util/struct_array.js

```diff
@@ -0,0 +1,334 @@
+ *  ]});
+ *
+ *  var pointArray = new PointArrayType();
+ *  pointArray.emplaceBack(10, 15);
+ *  pointArray.emplaceBack(20, 35);
+ *
+ *  point = pointArray.get(0);
+ *  assert(point.x === 10);
+ *  assert(point.y === 15);
+ *  point._setIndex(1);
+ *  assert(point.x === 20);
+ *  assert(point.y === 35);
+ *
+ * @private
+ */
+function StructArrayType(options) {
+
+    var key = JSON.stringify(options);
+    if (structArrayTypeCache[key]) {
+        return structArrayTypeCache[key];
+    }
+
+    if (options.alignment === undefined) options.alignment = 1;
+
+    function StructType() {
+        Struct.apply(this, arguments);
+    }
+
+    StructType.prototype = Object.create(Struct.prototype);
+
+    var offset = 0;
+    var maxSize = 0;
+    var usedTypes = ['Uint8'];
+
+    StructType.prototype.members = options.members.map(function(member) {
+        member = {
+            name: member.name,
+            type: member.type,
+            components: member.components || 1
+        };
+
+        assert(member.name.length);
+        assert(member.type in viewTypes);
+
+        if (usedTypes.indexOf(member.type) < 0) usedTypes.push(member.type);
+
+        var typeSize = sizeOf(member.type);
+        maxSize = Math.max(maxSize, typeSize);
+        member.offset = offset = align(offset, Math.max(options.alignment, typeSize));
+
+        for (var c = 0; c < member.components; c++) {
+            Object.defineProperty(StructType.prototype, member.name + (member.components === 1 ? '' : c), {
+                get: createGetter(member, c),
+                set: createSetter(member, c)
+            });
+        }
+
+        offset += typeSize * member.components;
+
+        return member;
+    });
+
+    StructType.prototype.alignment = options.alignment;
+    StructType.prototype.BYTE_SIZE = align(offset, Math.max(maxSize, options.alignment));
+
+    function StructArrayType() {
+        StructArray.apply(this, arguments);
+    }
+
+    StructArrayType.serialize = serializeStructArrayType;
+
+    StructArrayType.prototype = Object.create(StructArray.prototype);
+    StructArrayType.prototype.StructType = StructType;
+    StructArrayType.prototype.BYTES_PER_ELEMENT = StructType.prototype.BYTE_SIZE;
+    StructArrayType.prototype.emplaceBack = createEmplaceBack(StructType.prototype.members, StructType.prototype.BYTE_SIZE);
+    StructArrayType.prototype._usedTypes = usedTypes;
+
+
+    structArrayTypeCache[key] = StructArrayType;
+
+    return StructArrayType;
+}
+
+/**
+ * Serialize the StructArray type. This serializes the *type* not an instance of the type.
+ * @private
+ */
+function serializeStructArrayType() {
+    return {
+        members: this.prototype.StructType.prototype.members,
+        alignment: this.prototype.StructType.prototype.alignment,
+        BYTES_PER_ELEMENT: this.prototype.BYTES_PER_ELEMENT
+    };
+}
+
+
+function align(offset, size) {
+    return Math.ceil(offset / size) * size;
+}
+
+function sizeOf(type) {
```

Comment:
I didn't realize it was a protected symbol, I just capitalized it as if it were two words. `sizeOfType`? `sizeofType?`